enable plugins with transforms

frankieali opened the pull request:
Allowing remark to execute `parser.runSync()` so that plugin transforms can be applied to the AST tree. Demo here: https://frankieali.github.io/react-markdown/demo/dist/

Closes #188

frankieali commented:
I noticed that lists were not rendering as expected when I dropped [remark-toc](https://github.com/remarkjs/remark-toc) in for testing, since it was part of the original issue in #188. The renderer conditions were looking for `null` values but not `undefined` so I added them. It's not exactly related to `enabling plugins with transform` but it does relate to #188 so let me know if this should be separated out into a different PR.
I've updated my demo above with remark-TOC

callain commented:
@frankieali Will this also works for plugin like remark-gemoji-to-emoji ? If this is the case I might need your patch 👍

alifr commented:
@callain Yes, my demo worked fine when I dropped [remark-gemoji-to-emoji](https://github.com/jackycute/remark-gemoji-to-emoji) in.
I'm deep in another branch right now, but attached is a screen shot.
![gemoji-to-emoji-example](https://user-images.githubusercontent.com/5579274/58261196-dae27380-7d45-11e9-8a5c-5b4b7f7a5646.jpg)

frankieali commented:
@callain Yes, my demo worked fine when I dropped [remark-gemoji-to-emoji](https://github.com/jackycute/remark-gemoji-to-emoji) in.
I'm deep in another branch right now, but attached is a screen shot.

![gemoji-to-emoji-example](https://user-images.githubusercontent.com/20055089/58261332-17ae6a80-7d46-11e9-8b8e-457a5174ef26.jpg)

callain commented:
@frankieali Cool ! 
@rexxars Do you have a moment to check this PR and make a version if everything is ok ?

rexxars commented on src/ast-to-react.js:
```diff
@@ -2,12 +2,32 @@
+  if(node.position === undefined) {
+    visit(root, node, (nodeVisit, indexVisit, parentVisit) => {
+      // just in case parent does not have position either
+      const defaultPosition = {
```
Since defaultPosition is static, maybe hoist it to the top of the file?

rexxars commented on src/ast-to-react.js:
```diff
@@ -2,12 +2,32 @@
 function astToReact(node, options, parent = {}, index = 0) {
   const renderer = options.renderers[node.type]
 
+  if(node.type === 'root') {
```
Nitpick, but coding style wants a space between the if and the parentheses opening

rexxars commented on src/ast-to-react.js:
```diff
@@ -47,7 +67,7 @@ function getNodeProps(node, key, opts, renderer, parent, index) {
 
   if (opts.rawSourcePos && !isTagRenderer) {
-    props.sourcePosition = node.position
+    props.sourceposition = node.position
```
Was this accidental? Why was this changed?

frankieali commented:
I've made the requested updates. React was complaining once that `sourcePosition` should be `sourceposition` on my local, but I didn't look into it enough to understand why. I've reverted it back.

As far as the tests go, I'm seeing one failure on rawSourcePos, but it seems more like a fix to me since the rawSourcePos snapshot is undefined and now the test is returning a position object.

```
should pass on raw source position to non-tag renderers if rawSourcePos option is enabled

    expect(received).toMatchSnapshot()

    Snapshot name: `should pass on raw source position to non-tag renderers if rawSourcePos option is enabled 1`

    - Snapshot
    + Received

    - undefined
    + Position {
    +   "end": Object {
    +     "column": 6,
    +     "line": 1,
    +     "offset": 5,
    +   },
    +   "indent": Array [],
    +   "start": Object {
    +     "column": 1,
    +     "line": 1,
    +     "offset": 0,
    +   },
    + }
```

callain commented:
@rexxars Are the changes okay to you ?

rexxars commented:
There are still no tests for the new functionality - adding a minor plugin to the test suite to ensure that the transformation step is actually is executed would be nice. I'm not sure if there are good reasons why the nodes without a position needs one - if they do, maybe add a comment as to why, and include some tests to ensure that the position gets assigned as expected?

`root` is also assigned in the global module scope - I think modifying the `astToReact` signature a bit would be better, ensuring that the root is always explicit (the root-level node passed from `react-markdown.js`). Also not sure why it's assigned to an empty string by default, this seems like an odd choice.

callain commented:
@frankieali Hello, can you take @rexxars feedback in consideration ?

frankieali commented:
I've rebased my PR to include the latest in master. I've added some dev comments to the code and added the [Table of Contents](https://github.com/remarkjs/remark-toc) plugin as a dev-dependency for testing and demo purposes.

The reason for the `root` variable in the global module scope is to cache the root node so it can be walked later when trying to get position attributes from ancestor nodes. Code throughout the `ast-to-react.js` file make references to the `node.position` attribute. Tacking the parent's positioning onto child nodes seems the least obtrusive way to resolve this issue.

rexxars commented:
Thanks, I'll have a look!

simoncorry commented:
+1 for implementing this!

EricWVGG commented:
I gave this PR a try with about five different plugins last night and so far it's working fabulously.

cheshire137 commented:
This pull request is working for me to get the [gemoji](https://github.com/remarkjs/remark-gemoji) and [gemoji-to-emoji](https://github.com/jackycute/remark-gemoji-to-emoji) plugins working!

callain commented:
Hello @rexxars , hope you're good :) Can you review/merge this PR ? 😄

michelhelms-tdg commented:
Status? I would love to see that merged and the requested changes seem very minor.

rexxars commented:
Sorry, just moved to another country, been quite hectic. Will take a look at PRs this weekend.

rexxars commented:
Thanks so much for this! Released in 4.2.0
